exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/partitionsender/PartitionerDecorator.java

```diff
@@ -262,68 +280,122 @@ public FlushBatchesHandlingClass(boolean isLastBatch, boolean schemaChanged) {
-      try {
-        testCountDownLatch.await();
-      } catch (final InterruptedException e) {
-        logger.debug("Test only: partitioner thread is interrupted in test countdown latch await()", e);
-      }
-
-      final Thread currThread = Thread.currentThread();
-      final String currThreadName = currThread.getName();
-      final OperatorStats localStats = part.getStats();
-      try {
-        final String newThreadName = parentThreadName + currThread.getId();
-        currThread.setName(newThreadName);
+      final Thread thread = Thread.currentThread();
+      if (runner.compareAndSet(null, thread)) {
+        final String name = thread.getName();
+        thread.setName(String.format("Partitioner-%s-%d", partitionerDecorator.thread.getName(), thread.getId()));
+        final OperatorStats localStats = partitioner.getStats();
         localStats.clear();
         localStats.startProcessing();
-        iface.execute(part);
-      } catch (IOException e) {
-        exp = e;
-      } finally {
-        localStats.stopProcessing();
-        currThread.setName(currThreadName);
-        latch.countDown();
+        ExecutionException executionException = null;
+        try {
+          // Test only - Pause until interrupted by fragment thread
+          testCountDownLatch.await();
+          if (state.get() == STATE.NEW) {
+            iface.execute(partitioner);
+          }
+        } catch (InterruptedException e) {
+          if (state.compareAndSet(STATE.NEW, STATE.INTERRUPTED)) {
+            logger.warn("Partitioner Task interrupted during the run", e);
+          }
+        } catch (Throwable t) {
+          executionException = new ExecutionException(t);
+        } finally {
+          if (state.compareAndSet(STATE.NEW, STATE.COMPLETING)) {
+            if (executionException == null) {
+              localStats.stopProcessing();
+              state.lazySet(STATE.NORMAL);
+            } else {
+              exception = executionException;
+              state.lazySet(STATE.EXCEPTIONAL);
+            }
+          }
+          if (count.decrementAndGet() == 0) {
+            LockSupport.unpark(partitionerDecorator.thread);
+          }
+          thread.setName(name);
+          while (state.get() == STATE.INTERRUPTING) {
+            Thread.yield();
+          }
+          // Clear interrupt flag
+          Thread.interrupted();
+        }
+      }
+    }
+
+    void cancel(boolean mayInterruptIfRunning) {
+      Preconditions.checkState(Thread.currentThread() == partitionerDecorator.thread,
+          String.format("PartitionerTask can be cancelled only from the main %s thread", partitionerDecorator.thread.getName()));
+      if (runner.compareAndSet(null, partitionerDecorator.thread)) {
+        if (partitionerDecorator.executor instanceof ThreadPoolExecutor) {
+          ((ThreadPoolExecutor)partitionerDecorator.executor).remove(this);
+        }
+        count.decrementAndGet();
```

Comment:
The main PartitionDecorator thread may not get unparked.

1. Two PartitionerTasks are created.
2. Before any of the PartitionerTasks set their runner, the main Partitioner thread enters await() and cancels each PartitionerTask.
3. The cancel method executes ```
        if (partitionerDecorator.executor instanceof ThreadPoolExecutor) {
          ((ThreadPoolExecutor)partitionerDecorator.executor).remove(this);
        }
        count.decrementAndGet();```
    For each PartitionerTask.
 4. The count goes to zero, but the main PartitionerDecorator thread still calls Locksupport.park() in await().
 5. Unpark wasn't called when the count reached zero so the main PartitionDecorator thread can remain parked.